File: app/src/main/java/com/soundcloud/android/stream/StreamItem.java
package com.soundcloud.android.stream;


import com.google.auto.value.AutoValue;
import com.soundcloud.android.ads.AdData;
import com.soundcloud.android.ads.AppInstallAd;
import com.soundcloud.android.ads.VideoAd;
import com.soundcloud.android.api.model.ApiTrack;
import com.soundcloud.android.model.Urn;
import com.soundcloud.android.presentation.ListItem;
import com.soundcloud.android.suggestedcreators.SuggestedCreator;
import com.soundcloud.android.suggestedcreators.SuggestedCreatorItem;
import com.soundcloud.android.tracks.TrackItem;
import com.soundcloud.java.optional.Optional;

import java.util.ArrayList;
import java.util.List;

public abstract class StreamItem {

    public enum Kind {
        TRACK,
        PLAYLIST,
        FACEBOOK_LISTENER_INVITES,
        STATIONS_ONBOARDING,
        FACEBOOK_CREATORS,
        STREAM_UPSELL,
        SUGGESTED_CREATORS,
        APP_INSTALL,
        VIDEO_AD,
        STREAM_HIGHLIGHTS
    }

    static StreamItem forUpsell() {
        return new AutoValue_StreamItem_Default(Kind.STREAM_UPSELL);
    }

    public static StreamItem forSuggestedCreators(final List<SuggestedCreator> suggestedCreators) {
        return SuggestedCreators.create(suggestedCreators);
    }

    public static StreamItem forFacebookCreatorInvites(Urn trackUrn, String trackUrl) {
        return FacebookCreatorInvites.create(trackUrn, trackUrl);
    }

    public static StreamItem forFacebookListenerInvites() {
        return FacebookListenerInvites.create(Optional.absent());
    }

    public static FacebookListenerInvites forFacebookListenerInvites(final List<String> friendPictureUrls) {
        return FacebookListenerInvites.create(Optional.of(friendPictureUrls));
    }

    public static StreamItem forStationOnboarding() {
        return new AutoValue_StreamItem_Default(Kind.STATIONS_ONBOARDING);
    }

    public static StreamItem forAppInstall(AppInstallAd ad) {
        return AppInstall.create(ad);
    }

    public static StreamItem forVideoAd(VideoAd ad) {
        return Video.create(ad);
    }

    Optional<ListItem> getListItem() {
        if (kind() == Kind.TRACK) {
            return Optional.of(((TrackStreamItem) this).trackItem());
        } else if (kind() == Kind.PLAYLIST) {
            return Optional.of(((PlaylistStreamItem) this).playlistItem());
        }
        return Optional.absent();
    }

    public Optional<AdData> getAdData() {
        if (kind() == Kind.APP_INSTALL) {
            return Optional.of(((StreamItem.AppInstall) this).appInstall());
        } else if (kind() == Kind.VIDEO_AD) {
            return Optional.of(((StreamItem.Video) this).video());
        } else {
            return Optional.absent();
        }
    }

    public abstract Kind kind();

    public boolean isPromoted() {
        return (this.kind() == Kind.TRACK && ((TrackStreamItem) this).promoted())
                || (this.kind() == Kind.PLAYLIST && ((PlaylistStreamItem) this).promoted());
    }

    public boolean isAd() {
        return kind() == Kind.APP_INSTALL || kind() == Kind.VIDEO_AD;
    }

    public boolean isUpsell() {
        return kind() == Kind.STREAM_UPSELL;
    }

    @AutoValue
    abstract static class Default extends StreamItem {
    }

    @AutoValue
    public abstract static class FacebookCreatorInvites extends StreamItem {
        public abstract Urn trackUrn();

        abstract String trackUrl();

        private static FacebookCreatorInvites create(Urn trackUrn, String trackUrl) {
            return new AutoValue_StreamItem_FacebookCreatorInvites(Kind.FACEBOOK_CREATORS, trackUrn, trackUrl);
        }
    }

    @AutoValue
    public abstract static class FacebookListenerInvites extends StreamItem {
        public abstract Optional<List<String>> friendPictureUrls();

        public boolean hasPictures() {
            return friendPictureUrls().isPresent() && !friendPictureUrls().get().isEmpty();
        }

        private static FacebookListenerInvites create(Optional<List<String>> friendPictureUrls) {
            return new AutoValue_StreamItem_FacebookListenerInvites(Kind.FACEBOOK_LISTENER_INVITES, friendPictureUrls);
        }
    }

    @AutoValue
    public abstract static class SuggestedCreators extends StreamItem {
        public abstract List<SuggestedCreatorItem> suggestedCreators();

        public static SuggestedCreators create(List<SuggestedCreator> suggestedCreators) {
            final List<SuggestedCreatorItem> suggestedCreatorItems = new ArrayList<>(suggestedCreators.size());
            for (SuggestedCreator suggestedCreator : suggestedCreators) {
                suggestedCreatorItems.add(SuggestedCreatorItem.fromSuggestedCreator(suggestedCreator));
            }
            return new AutoValue_StreamItem_SuggestedCreators(Kind.SUGGESTED_CREATORS, suggestedCreatorItems);
        }
    }

    @AutoValue
    public abstract static class StreamHighlights extends StreamItem {
        public abstract List<TrackItem> suggestedTrackItems();

        public static StreamHighlights create(List<ApiTrack> suggestedTracks) {
            final List<TrackItem> suggestedTrackItems = new ArrayList<>(suggestedTracks.size());
            for (ApiTrack apiTrack : suggestedTracks) {
                suggestedTrackItems.add(TrackItem.from(apiTrack));
            }
            return new AutoValue_StreamItem_StreamHighlights(Kind.STREAM_HIGHLIGHTS, suggestedTrackItems);
        }
    }

    @AutoValue
    public abstract static class AppInstall extends StreamItem {
        public abstract AppInstallAd appInstall();

        static AppInstall create(AppInstallAd ad) {
            return new AutoValue_StreamItem_AppInstall(Kind.APP_INSTALL, ad);
        }
    }

    @AutoValue
    public abstract static class Video extends StreamItem {
        public abstract VideoAd video();

        static Video create(VideoAd ad) {
            return new AutoValue_StreamItem_Video(Kind.VIDEO_AD, ad);
        }
    }
}

File: app/src/main/java/com/soundcloud/android/stream/StreamItem.kt
package com.soundcloud.android.stream

import com.soundcloud.android.ads.AdData
import com.soundcloud.android.ads.AppInstallAd
import com.soundcloud.android.ads.VideoAd
import com.soundcloud.android.events.CurrentPlayQueueItemEvent
import com.soundcloud.android.events.LikesStatusEvent
import com.soundcloud.android.events.RepostsStatusEvent
import com.soundcloud.android.model.Urn
import com.soundcloud.android.playlists.Playlist
import com.soundcloud.android.playlists.PlaylistItem
import com.soundcloud.android.presentation.LikeableItem
import com.soundcloud.android.presentation.PlayableItem
import com.soundcloud.android.presentation.RepostableItem
import com.soundcloud.android.presentation.UpdatablePlaylistItem
import com.soundcloud.android.presentation.UpdatableTrackItem
import com.soundcloud.android.suggestedcreators.SuggestedCreator
import com.soundcloud.android.suggestedcreators.SuggestedCreatorItem
import com.soundcloud.android.tracks.Track
import com.soundcloud.android.tracks.TrackItem
import com.soundcloud.android.view.adapters.PlayableViewItem
import com.soundcloud.java.optional.Optional
import java.util.Date

sealed class StreamItem(val kind: Kind) {

    internal val playableItem: Optional<PlayableItem>
        get() = when (this) {
            is TrackStreamItem -> Optional.of(this.trackItem)
            is PlaylistStreamItem -> Optional.of(this.playlistItem)
            else -> Optional.absent()
        }

    val adData: Optional<AdData>
        get() = when (this) {
            is AppInstall -> Optional.of(this.appInstallAd)
            is Video -> Optional.of(this.video)
            else -> Optional.absent()
        }

    val isPromoted: Boolean
        get() = when (this) {
            is TrackStreamItem -> this.promoted
            is PlaylistStreamItem -> this.promoted
            else -> false
        }
    val isAd: Boolean
        get() = this is AppInstall || this is Video

    val isUpsell: Boolean
        get() = this is Upsell

    enum class Kind {
        TRACK,
        PLAYLIST,
        FACEBOOK_LISTENER_INVITES,
        FACEBOOK_CREATORS,
        STREAM_UPSELL,
        SUGGESTED_CREATORS,
        APP_INSTALL,
        VIDEO_AD
    }

    data class FacebookCreatorInvites(val trackUrn: Urn, val trackUrl: String) : StreamItem(Kind.FACEBOOK_CREATORS)

    data class FacebookListenerInvites(val friendPictureUrls: Optional<List<String>> = Optional.absent()) : StreamItem(Kind.FACEBOOK_LISTENER_INVITES) {
        fun hasPictures(): Boolean {
            return friendPictureUrls.isPresent && !friendPictureUrls.get().isEmpty()
        }
    }

    data class SuggestedCreators(val suggestedCreators: List<SuggestedCreatorItem>) : StreamItem(Kind.SUGGESTED_CREATORS) {
        companion object {

            fun create(suggestedCreators: List<SuggestedCreator>): SuggestedCreators {
                return SuggestedCreators(suggestedCreators.map { SuggestedCreatorItem.fromSuggestedCreator(it) })
            }
        }
    }

    data class AppInstall(val appInstallAd: AppInstallAd) : StreamItem(Kind.APP_INSTALL)

    data class Video(val video: VideoAd) : StreamItem(Kind.VIDEO_AD)

    object Upsell : StreamItem(Kind.STREAM_UPSELL)
}

data class TrackStreamItem(val trackItem: TrackItem, val promoted: Boolean, val createdAt: Date, val avatarUrlTemplate: Optional<String>) : StreamItem(Kind.TRACK),
                                                                                                                                            PlayableViewItem<TrackStreamItem>,
                                                                                                                                            UpdatableTrackItem,
                                                                                                                                            LikeableItem,
                                                                                                                                            RepostableItem {

    override val urn: Urn
        get() = trackItem.urn

    override fun updatedWithTrack(track: Track): TrackStreamItem {
        return copy(trackItem = trackItem.updatedWithTrack(track))
    }

    override fun updatedWithLike(likeStatus: LikesStatusEvent.LikeStatus): TrackStreamItem {
        return copy(trackItem = trackItem.updatedWithLike(likeStatus))
    }

    override fun updatedWithRepost(repostStatus: RepostsStatusEvent.RepostStatus): TrackStreamItem {
        return copy(trackItem = trackItem.updatedWithRepost(repostStatus))
    }

    override fun updateNowPlaying(event: CurrentPlayQueueItemEvent): TrackStreamItem {
        val updatedTrackItem = trackItem.updateNowPlaying(event.currentPlayQueueItem.urnOrNotSet)
        return copy(trackItem = updatedTrackItem)
    }

    companion object {

        fun create(trackItem: TrackItem, createdAt: Date, avatarUrlTemplate: Optional<String>): TrackStreamItem {
            return TrackStreamItem(trackItem, trackItem.isPromoted, createdAt, avatarUrlTemplate)
        }
    }
}

data class PlaylistStreamItem(val playlistItem: PlaylistItem, val promoted: Boolean, val createdAt: Date, val avatarUrlTemplate: Optional<String>) : StreamItem(Kind.PLAYLIST),
                                                                                                                                                     LikeableItem,
                                                                                                                                                     RepostableItem,
                                                                                                                                                     UpdatablePlaylistItem {

    override val urn: Urn
        get() = playlistItem.urn

    override fun updatedWithLike(likeStatus: LikesStatusEvent.LikeStatus): PlaylistStreamItem {
        return copy(playlistItem = playlistItem.updatedWithLike(likeStatus))
    }

    override fun updatedWithRepost(repostStatus: RepostsStatusEvent.RepostStatus): PlaylistStreamItem {
        return copy(playlistItem = playlistItem.updatedWithRepost(repostStatus))
    }

    override fun updatedWithPlaylist(playlist: Playlist): PlaylistStreamItem {
        return copy(playlistItem = playlistItem.toBuilder().playlist(playlist).build())
    }

    override fun updatedWithTrackCount(trackCount: Int): PlaylistStreamItem {
        return copy(playlistItem = playlistItem.updatedWithTrackCount(trackCount))
    }

    override fun updatedWithMarkedForOffline(markedForOffline: Boolean): PlaylistStreamItem {
        return copy(playlistItem = playlistItem.updatedWithMarkedForOffline(markedForOffline))
    }

    companion object {

        internal fun create(playlistItem: PlaylistItem,
                            createdAt: Date, avatarUrlTemplate: Optional<String>): PlaylistStreamItem {
            return PlaylistStreamItem(playlistItem, playlistItem.isPromoted, createdAt, avatarUrlTemplate)
        }
    }
}
